ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,388 @@
+from rcl_interfaces.msg import ParameterDescriptor, ParameterType
+from sensor_msgs.msg import CompressedImage, CameraInfo
+from std_srvs.srv import SetBool
+
+# Local imports
+from ada_feeding_msgs.msg import FaceDetection
+from ada_feeding_perception.helpers import (
+    download_checkpoint,
+    get_img_msg_type,
+    cv2_image_to_ros_msg,
+)
+
+
+class FaceDetectionNode(Node):
+    """
+    This node publishes a 3d PointStamped location
+    of the largest detected mouth with respect to camera_depth_optical_frame.
+    """
+
+    # pylint: disable=too-many-instance-attributes
+    # Needed for multiple model loads, publisher, subscribers, and shared variables
+    def __init__(
+        self,
+    ):
+        """
+        Initializes the FaceDetection node. This node exposes a toggle_face_detection
+        service that can be used to toggle the face detection on or off and
+        publishes information about detected faces to the /face_detection
+        topic when face detection is on.
+        """
+        super().__init__("face_detection")
+
+        # Read the parameters
+        # NOTE: These parameters are only read once. Any changes after the node
+        # is initialized will not be reflected.
+        (
+            face_model_name,
+            face_model_base_url,
+            landmark_model_name,
+            landmark_model_base_url,
+            model_dir,
+        ) = self.read_params()
+        self.face_model_name = face_model_name.value
+        self.landmark_model_name = landmark_model_name.value
+
+        # Download the checkpoints if they don't exist
+        self.face_model_path = os.path.join(model_dir.value, self.face_model_name)
+        if not os.path.isfile(self.face_model_path):
+            self.get_logger().info(
+                "Face detection model checkpoint does not exist. Downloading..."
+            )
+            download_checkpoint(
+                self.face_model_name, model_dir.value, face_model_base_url.value
+            )
+            self.get_logger().info(
+                f"Model checkpoint downloaded {self.face_model_path}."
+            )
+        self.landmark_model_path = os.path.join(
+            model_dir.value, self.landmark_model_name
+        )
+        if not os.path.isfile(self.landmark_model_path):
+            self.get_logger().info(
+                "Facial landmark model checkpoint does not exist. Downloading..."
+            )
+            download_checkpoint(
+                self.landmark_model_name, model_dir.value, landmark_model_base_url.value
+            )
+            self.get_logger().info(
+                f"Model checkpoint downloaded {self.landmark_model_path}."
+            )
+
+        # Keeps track of whether face detection is on or not
+
+        self.is_on = False
+        self.is_on_lock = Lock()
+        self.bridge = CvBridge()
+
+        self.is_face_detected = False
+        self.depth_buffer = collections.deque(maxlen=10)
```

Comment:
Also, can you make the length of `self.depth_buffer` a parameter with default value 10? Because if in the future we increase the FPS of the depth camera, the number of images stored in the buffer will also need to be increased proportionally.